## Examples

This folder contains iPython notebooks for examples of Rosely data analysis. The notebooks may take a few seconds to open, be patient. If your browser does not open the notebooks, try Firefox or Google Chrome. If GitHub is too busy to open it, download this repo and open the .html files in your browser.

### RNA-seq data analysis with paired samples (full explanations, important to read first!!)
For a normal RNA-seq analysis (paired), including identification of differential gene expression, Gene Ontology and KEGG pathway analysis, see file [human_classical_vs_nonclassical_monocyte_study_RNA-seq_analysis.ipynb](https://github.com/yjiangnan/rosely/blob/master/examples/human_classical_vs_nonclassical_monocyte_study_RNA-seq_analysis.ipynb). It also shows how neutrality-controlled *p*-values are used for false discovery control. 

### RNA-seq data analysis with unpaired samples
For a normal RNA-seq analysis (unpaired), including identification of differential gene expression, Gene Ontology and KEGG pathway analysis, see file [Hematopoietic_stem_cells.RNA-seq.ipynb](https://github.com/yjiangnan/rosely/blob/master/examples/Hematopoietic_stem_cells.RNA-seq.ipynb). It also shows how Rosely can identify experimental design problems even the researchers themselves failed to recognize. 

### Code for reproducing manuscript results
The folder [manuscript_code](manuscript_code) contains iPython notebooks to reproduce the results in the munuscript entitled "**High-throughput data analysis by Rosely: a genomic approach to the reproducibility crisis**".
